### PR TITLE
New check 89: check documentation (lines of)

### PR DESCRIPTION
**Pull request by se38**

### CODE REVIEW

**Comment by se38** [comment]
abaplint "All checks have passed" :-)

File: src/checks/zcl_aoc_super.clas.abap [hunk 1]
```diff
@@ -71,6 +71,9 @@ CLASS zcl_aoc_super DEFINITION
         !it_tokens       TYPE stokesx_tab
       RETURNING
         VALUE(rv_result) TYPE token_row .
+    CLASS-METHODS get_destination
+      IMPORTING iv_srcid         TYPE scr_source_id
```
Comment (larshp) [inline src/checks/zcl_aoc_super.clas.abap]: I guess there is no need to import SCRID, as the superclass knows about it?

**Comment by se38** [comment]
There's indeed no need. But it's more a philosophical question ;-) I don't like global vars. I pass them to the methods just for dependency documentation.

But I'll change it though and come back later today.

**Comment by larshp** [comment]
I'm just worried that SCRID does not exist in lower releases, so would like to access it centrally

**Comment by se38** [comment]
You where absolutely right: field SRCID was not valid in 7.02 (just checked) -> fixed

File: src/checks/zcl_aoc_check_89.clas.abap [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+    mv_minlength = 10.
+
+  ENDMETHOD.                    "CONSTRUCTOR
+
+  METHOD get_message_text.
+
+    CLEAR p_text.
+
+    IF p_code IS NOT INITIAL.
+      p_text = |{ 'Minimum lines of documentation not reached:'(m01) } { p_code }|.
```
Comment (pcf0) [inline src/checks/zcl_aoc_check_89.clas.abap]: If you use inform with `p_param_1` and `p_param_2` you could use this code:
```abap
    CASE p_code.
      WHEN '001'.
        p_text = 'Minimum lines of documentation not reached: &1 < &2'(m01) .
      WHEN OTHERS.
        super->get_message_text( EXPORTING p_test = p_test
                                           p_code = p_code
                                 IMPORTING p_text = p_text ).
    ENDCASE.
```

File: src/checks/zcl_aoc_check_89.clas.abap [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+
+      READ TABLE it_levels
+        WITH KEY level = 0
+        ASSIGNING <ls_level>.
+
+      inform( p_sub_obj_type = c_type_include
+              p_sub_obj_name = <ls_level>-name
+              p_kind         = mv_errty
+              p_test         = myname
+              p_code         = |{ lv_actual_length } < { mv_minlength }| ).
```
Comment (pcf0) [inline src/checks/zcl_aoc_check_89.clas.abap]: Should `p_code` not be constant (e.g. `001`) and instead `p_param_1` and `p_param_2` filled?

**Comment by se38** [comment]
I know the solution with parameters (like check 70). 
Problem: if I change the logic of check 89 to this solution, in Eclipse we don't see the message but just the check number. You can see the difference here in this picture (changed check 89, a coming new check 90 with my logic, and the already existant check 70)
adt_atc

**Comment by larshp** [comment]
hmm, so this is a general problem?

**Comment by larshp** [comment]
I get the full message text, text from check 70 selected,
image

**Comment by se38** [comment]
The standard checks are okay. Maybe it's because of the logon language DE. Will check later.

**Comment by pcf0** [comment]
I just set up eclipse for the first time to test this. I notice the same problem.
I set the language in SU01 to EN and used logon language EN.
The standard checks are okay. The abapOpenChecks are not okay.
grafik

But this seems very random. I checked the class `ZCL_ABAPGIT_APACK_MIGRATION` and added this "TODO", too.
The most checks are okay, but "Find comment markers" shows only "001":
grafik

**Comment by larshp** [comment]
I think we should try identifying the issue, and fix all checks so that they work similarly when issuing messages

**Comment by se38** [comment]
Yes, this was my plan for the coming week

**Comment by se38** [comment]
Got the problem (also documented in "How to write an ATC Check" on SCN).
In the constructor you have to add the message in the instance attribute "scimessages":
```
    ls_scimessage-test = myname.
    ls_scimessage-code = '001'.
    ls_scimessage-kind = c_error.
    ls_scimessage-text = |{ 'Minimum lines of documentation not reached:'(m01) } &1|.
    ls_scimessage-pcom = ''.
    ls_scimessage-pcom_alt = ''.

    INSERT ls_scimessage INTO TABLE scimessages.
```

The parameter "&1" will be replaced in method CHECK with p_param_1.
```
      inform( p_sub_obj_type = c_type_include
              p_sub_obj_name = <ls_level>-name
              p_kind         = mv_errty
              p_test         = myname
              p_code         = '001'
              p_param_1  = |{ lv_actual_length } < { mv_minlength }| ).
```

We don't need the method "get_message_text".
Updated my fork.

**Comment by larshp** [comment]
cool, I'd suggest adding a helper method for adding rows in `scimessages` but we can add that later

looking at the diff, https://github.com/larshp/abapOpenChecks/pull/692/files, there are changes to check 40, this seems strange

**Comment by se38** [comment]
I updated my fork with the most current versions of OpenChecks today. That's why the updated check 40 is also in my commit. I have to work out how to prevent this behaviour. I'm still a Git newby ...

The check 70 will work too now. Will push the changes after check 89 and 90.
check_70

**Comment by larshp** [comment]
will try merging it when I have some time, need to double check what happens with check 40

**Comment by se38** [comment]
As I've said earlier: the "check 40" in my pull request is just the refresh from your master. Code should be the same.

If it's easier for you: we can close this pull request and create a new one with only my changes (check 89 and SUPER).

**Comment by larshp** [comment]
nah, I'll try merging when I have the time

if its just a refresh from master it typically does not show up in https://github.com/larshp/abapOpenChecks/pull/692/files as it does in this case 🤷‍♂

File: src/checks/zcl_aoc_check_89.clas.abap [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+      READ TABLE it_levels
+        WITH KEY level = 0
+        ASSIGNING <ls_level>.
+
+      inform( p_sub_obj_type = c_type_include
+              p_sub_obj_name = <ls_level>-name
+              p_kind         = mv_errty
+              p_test         = myname
+              p_code         = '001'
+              p_param_1  = |{ lv_actual_length } < { mv_minlength }| ).
```
Comment (pcf0) [inline src/checks/zcl_aoc_check_89.clas.abap]: As I said earlier, i would use
```suggestion
              p_param_1  = |{ lv_actual_length }|
              p_param_2  = |{ mv_minlength }|.
```

File: src/checks/zcl_aoc_check_89.clas.abap [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+
+    enable_rfc( ).
+
+    mv_errty = c_error.
+    mv_minlength = 10.
+
+    ls_scimessage-test = myname.
+    ls_scimessage-code = '001'.
+    ls_scimessage-kind = c_error.
+    ls_scimessage-text = |{ 'Minimum lines of documentation not reached:'(m01) } &1|.
```
Comment (pcf0) [inline src/checks/zcl_aoc_check_89.clas.abap]: I would always include wildcards in translatable text. This is the only way to also support RTL languages or languages with different syntax.
So in combination with my other suggestion:
```suggestion
    ls_scimessage-text = 'Minimum lines of documentation not reached: &1 < &2'(m01).
```

File: src/checks/zcl_aoc_check_89.clas.xml [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+    <FIXPT>X</FIXPT>
+    <UNICODE>X</UNICODE>
+    <RSTAT>P</RSTAT>
+    <WITH_UNIT_TESTS>X</WITH_UNIT_TESTS>
+   </VSEOCLASS>
+   <TPOOL>
+    <item>
+     <ID>I</ID>
+     <KEY>M01</KEY>
+     <ENTRY>Minimum lines of documentation not reached:</ENTRY>
```
Comment (pcf0) [inline src/checks/zcl_aoc_check_89.clas.xml]: So this should also be:
```suggestion
     <ENTRY>Minimum lines of documentation not reached: &1 &lt; &2</ENTRY>
```

**Comment by larshp** [comment]
I agree with the comments from @pcf0 above, but will try merging anyhow, we can fix above in follow up PR

**Comment by larshp** [comment]
opened https://github.com/larshp/abapOpenChecks/issues/697 for above comments

**Pull request merged**